### PR TITLE
wip

### LINKED ISSUE #1: Requirements :P
I have no idea what you are trying to accomplish here but I have a small ZIO-AWS wrapper myself, so let me list off what I have needed :)

- The AWS SDK reads implicit credentials, which is essentially never what you want. Make sure a client is passed credentials
- You can fetch S3 files onto disk or as an `OutputStream`. The latter is very useful for larger files
- There is a `Region` class but you can make invalid instances of it, which will blow things up with runtime exceptions

I solved the `Region` issue like so:

```scala
def region(r: String): Option[Region] =
    Option(Region.of(r)).filter(Region.regions.contains(_))
```

It would be sweet if the S3 wrapper could stream me file contents line-by-line (i.e. for CSV or line-break separated JSONs).

### LINKED ISSUE #4: release snapshot
dmijicToday at 4:10 PM
@here Just released snapshots for the following projects:

- zio-interop-cats
- zio-interop-scalaz
- zio-interop-monix
- zio-interop-twitter
- zio-interop-reactivestreams
- zio-interop-java

As there are quite a few projects in the organization nowadays, I'd like to ask maintainers of projects that has release process enabled to help @softinio and me by aligning their projects with the latest changes. The following PR can be used as an example: https://github.com/zio/interop-twitter/pull/75/files.

### PR DESCRIPTION
**Pull request by regis-leray**
- [x] unit test
- [x] run integration test with amazon s3
- [x] Provide Test Environment
- [ ] Documentation
- [x] update ssh key for github-pages
- [x] Review from ZIO contributors

#closes #1 
#closes #4

### CODE REVIEW

**Comment by softinio** [comment]
@regis-leray here you go as discussed please replace `fingerprint` for microsite step of circleci config to this please: `b0:70:84:70:c9:31:62:54:1f:29:49:4d:22:4d:e2:eb`

Thanks anything else I can help with let me know.

File: build.sbt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  .in(file("."))
+  .settings(
+    skip in publish := true,
+    libraryDependencies ++= Seq(
+      "dev.zio"                %% "zio"                         % "1.0.0-RC17",
+      "dev.zio"                %% "zio-streams"                 % "1.0.0-RC17",
+      "dev.zio"                %% "zio-nio"                     % "0.4.0",
+      "dev.zio"                %% "zio-interop-java"            % "1.1.0.0-RC6",
+      "dev.zio"                %% "zio-interop-reactivestreams" % "1.0.3.5-RC2",
+      "software.amazon.awssdk" % "s3"                           % "2.10.50",
```
Comment (NeQuissimus) [inline build.sbt]: If we ever add scala-steward to this, the AWS dependency is going to be a giant pain :D May want to include mergify at that point and auto-merge scala-steward PRs :D

**Comment by jdegoes** [comment]
@regis-leray Awesome work, can we schedule a time to review live over video?

**Comment by regis-leray** [comment]
@jdegoes of course :)

File: src/main/scala/zio/s3/settings.scala [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import java.net.URI
+
+import software.amazon.awssdk.regions.Region
+import zio.{ IO, ZIO }
+
+import scala.util.Try
+
+final case class S3Credentials(accessKeyId: String, secretAccessKey: String)
+
+final private[s3] case class S3Settings(region: Region, credentials: S3Credentials, uriEndpoint: Option[URI])
```
Comment (dmytroorlov-kreditech) [inline src/main/scala/zio/s3/settings.scala]: when `S3Settings` is private I cannot write in DIstage:
```
make[S3.Service[Any]].fromResource(zio.s3.S3.Live.connect(_: S3Settings))
```

File: src/main/scala/zio/s3/errors.scala [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.s3
+
+import java.io.IOException
+
+sealed trait S3Failure
```
Comment (dmytroorlov-kreditech) [inline src/main/scala/zio/s3/errors.scala]: with
```
sealed trait S3Failure {
  def message: String

  def cause: Throwable
}
```
I can improve my errors with `cause`:
```
def s3Exception(cause: S3Exception) = new RuntimeException(s"s3Exception: $cause", cause)
```

File: src/main/scala/zio/s3/settings.scala (hunk 2 above)
Comment (regis-leray) [reply src/main/scala/zio/s3/settings.scala]: thanks for your early comments. I will try to fix it

File: src/main/scala/zio/s3/errors.scala (hunk 3 above)
Comment (regis-leray) [reply src/main/scala/zio/s3/errors.scala]: S3Exception dont extend S3Failure, 
im only exposing the error from amazon sdk `software.amazon.awssdk.services.s3.model.S3Exception`

i should rename S3Failure  to S3ConnectionFailure to remove any ambiguity

Comment (dmytroorlov-kreditech) [reply src/main/scala/zio/s3/errors.scala]: sorry, wrong copied line, should be
```
    def s3Err(e: S3Failure) = new RuntimeException(s"s3Err: $e", e.cause)
```

File: build.sbt (hunk 1 above)
Comment (regis-leray) [reply build.sbt]: I would be great if we had a offcial zio http client. I could replace this dependency

File: src/main/scala/zio/s3/settings.scala (hunk 2 above)
Comment (regis-leray) [reply src/main/scala/zio/s3/settings.scala]:
```
val f: S3Settings => Managed[S3ConnectionFailure, S3] = S3.Live.connect(_: S3Settings)
```

this compile fine, i dont really understand, maybe it is a distage problem, i never use it

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,264 @@
+version: 2.1
+
+scala_212: &scala_212
```
Comment (mijicd) [inline .circleci/config.yml]: Scala versions can be bumped.

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+        sudo apt update
+        sudo apt install openjdk-${JDK_VERSION}-jdk -y
+
+load_cache: &load_cache
+  - restore_cache:
+      key: sbt-nio-cache
+
+save_cache: &save_cache
+  - save_cache:
+      key: sbt-nio-cache
```
Comment (mijicd) [inline .circleci/config.yml]: Change cache key.

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+        [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
+        nvm install
+        nvm use
+        node -v
+
+microsite: &microsite
+  steps:
+    - add_ssh_keys:
+        fingerprints:
+          - "b0:70:84:70:c9:31:62:54:1f:29:49:4d:22:4d:e2:eb"
```
Comment (mijicd) [inline .circleci/config.yml]: Make sure to have a valid fingerprint once microsite is ready to be released.

File: build.sbt (hunk 1 above)
Comment (mijicd) [reply build.sbt]: @regis-leray Can you go with sttp until the client is ready?

@NeQuissimus I think we need mergify throughout the whole organization to take care of steward stuff :).

File: src/main/scala/zio/s3/S3.scala [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+object S3 {
+
+  trait Service[R] {
+    def createBucket(bucketName: String): ZIO[R, S3Exception, Unit]
+
+    def deleteBucket(bucketName: String): ZIO[R, S3Exception, Unit]
+
+    def isBucketExists(bucketName: String): ZIO[R, S3Exception, Boolean]
+
+    def listBuckets(): ZIO[R, S3Exception, S3BucketListing]
```
Comment (mijicd) [inline src/main/scala/zio/s3/S3.scala]: You can drop the empty parens here.

File: src/main/scala/zio/s3/S3.scala [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+        .map(client =>
+          new S3 {
+            val s3 = new Live(client)
+          }
+        )
+        .mapError(e => ConnectionError(e.getMessage, e.getCause))
+  }
+
+  //
+  final class Test(path: ZPath) extends Service[Blocking] {
```
Comment (mijicd) [inline src/main/scala/zio/s3/S3.scala]: I'd pull this one out.

File: src/main/scala/zio/s3/S3.scala [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+//  object Test {
+//
+//    def connect(path: Path): ZManaged[Blocking, Nothing, S3] =
+//      ZManaged.succeed(new S3 {
+//        protected override val s3: Service[Any] = new Test(ZPath.fromJava(path))
+//      })
+//  }
+}
+
+case class S3BucketListing(buckets: List[S3Bucket])
```
Comment (mijicd) [inline src/main/scala/zio/s3/S3.scala]: Maybe pull them out in separate files or to the companion, and remove `S3` prefix.

File: src/main/scala/zio/s3/ZIOUtils.scala [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+ */
+package zio.s3
+
+import java.nio.ByteBuffer
+
+import zio.stream.ZStream
+import zio.{ Chunk, ZIO }
+
+/*
+Need to be removed when ZIO-RC18 is released
```
Comment (mijicd) [inline src/main/scala/zio/s3/ZIOUtils.scala]: If you like to "live on the edge" you can depend on snapshot :).

File: src/main/scala/zio/s3/S3.scala [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+import zio.blocking.Blocking
+import zio.interop.javaz.fromCompletionStage
+import zio.interop.reactiveStreams._
+import zio.nio.file.{ Path => ZPath, _ }
+import zio.stream.{ ZSink, ZStream, ZStreamChunk }
+import zio._
+import scala.jdk.CollectionConverters._
+
+trait S3 {
+  val s3: S3.Service[Any]
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]:
```scala
type S3 = Has[S3.Service]
```

File: src/main/scala/zio/s3/S3.scala [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+}
+
+object S3 {
+
+  trait Service[R] {
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]: The `R` type parameter goes away completely.

File: src/main/scala/zio/s3/S3.scala [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+    def createBucket(bucketName: String): ZIO[R, S3Exception, Unit]
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]:
```suggestion
    def createBucket(bucketName: String): IO[S3Exception, Unit]
```

File: src/main/scala/zio/s3/S3.scala [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+      bucketName: String,
+      key: String,
+      contentType: String,
+      content: ZStream[R1, Throwable, Byte]
+    ): ZIO[R1, S3Exception, Unit]
+
+    def execute[T](f: S3AsyncClient => CompletableFuture[T]): ZIO[R, S3Exception, T]
+  }
+
+  final class Live private (unsafeClient: S3AsyncClient) extends Service[Any] {
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]:
```scala
def live(unsafeClient: S3AsyncClient): Layer[Nothing, S3] = ...
```

File: src/main/scala/zio/s3/S3.scala [hunk 15]
```diff
@@ -0,0 +1,395 @@
...
+          ).map(S3ObjectListing(_))
+        }
+
+    override def putObject[R1](
+      bucketName: String,
+      key: String,
+      contentLength: Long,
+      contentType: String,
+      content: ZStreamChunk[R1, Throwable, Byte]
+    ): ZIO[Any with R1, S3Exception, Unit] =
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]:
```suggestion
    ): ZIO[R1, S3Exception, Unit] =
```

File: src/main/scala/zio/s3/S3.scala [hunk 16]
```diff
@@ -0,0 +1,395 @@
...
+          )
+        )
+        .unit
+
+    // only for file which are bigger than 5 Mb
+    def multipartUpload[R1 <: Any](n: Int)(
+      bucketName: String,
+      key: String,
+      contentType: String,
+      content: ZStream[R1, Throwable, Byte]
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]: Change to `ZStreamChunk` of byte to minimize overhead.

File: src/main/scala/zio/s3/S3.scala [hunk 17]
```diff
@@ -0,0 +1,395 @@
...
+                         .build()
+                     )
+                   ).map(_.uploadId())
+
+        parts <- content
+                //part size limit is 5Mb, required by amazon api
+                  .chunkN(5 * 1024 * 1024)
+                  .chunks
+                  .zipWithIndex
+                  .mapMPar(n) {
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]: TODO: Talk to Itamar: 

```suggestion
                  .mapM {
```

File: src/main/scala/zio/s3/S3.scala [hunk 18]
```diff
@@ -0,0 +1,395 @@
...
+                  .key(key)
+                  .multipartUpload(CompletedMultipartUpload.builder().parts(parts.asJavaCollection).build())
+                  .uploadId(uploadId)
+                  .build()
+              )
+            )
+      } yield ()
+
+    def execute[T](f: S3AsyncClient => CompletableFuture[T]): ZIO[Any, S3Exception, T] =
+      fromCompletionStage(UIO.apply(f(unsafeClient))).refineToOrDie[S3Exception]
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]: This might be inside ZIO now.

File: src/main/scala/zio/s3/S3.scala [hunk 19]
```diff
@@ -0,0 +1,395 @@
...
+
+    override def exceptionOccurred(error: Throwable): Unit = {
+      cf.completeExceptionally(error)
+      ()
+    }
+  }
+
+  object Live {
+
+    def connect(region: Region, credentials: S3Credentials): Managed[ConnectionError, S3] =
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]:
```suggestion
    def connect(region: Region, credentials: S3Credentials): Layer[ConnectionError, S3] =
```

Move outside of `Live` and just put inside companion object of S3.

File: src/main/scala/zio/s3/S3.scala [hunk 20]
```diff
@@ -0,0 +1,395 @@
...
+      for {
+        settings <- S3Settings
+                     .from(region, credentials)
+                     .toManaged_
+                     .mapError(e => ConnectionError(e.getMessage, e.getCause))
+        client <- connect(settings, None)
+      } yield client
+
+    def connect(settings: S3Settings, uriEndpoint: Option[URI]): Managed[ConnectionError, S3] =
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]: Same here

File: src/main/scala/zio/s3/S3.scala (hunk 8 above)
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]:
```scala
def test(path: ZPath): ZLayer[Blocking, Nothing, S3]
```

File: src/main/scala/zio/s3/S3.scala (hunk 9 above)
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]: Use a `List[S3Bucket]` instead.

File: src/main/scala/zio/s3/S3.scala [hunk 21]
```diff
@@ -0,0 +1,395 @@
...
+
+case class S3BucketListing(buckets: List[S3Bucket])
+
+object S3BucketListing {
+
+  def apply(resp: ListBucketsResponse): S3BucketListing =
+    S3BucketListing(resp.buckets().asScala.toList.map(S3Bucket(_)))
+}
+
+case class S3Bucket(name: String, creationDate: Instant)
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]:
```suggestion
final case class S3Bucket(name: String, creationDate: Instant)
```

File: src/main/scala/zio/s3/S3.scala [hunk 22]
```diff
@@ -0,0 +1,395 @@
...
+//        protected override val s3: Service[Any] = new Test(ZPath.fromJava(path))
+//      })
+//  }
+}
+
+case class S3BucketListing(buckets: List[S3Bucket])
+
+object S3BucketListing {
+
+  def apply(resp: ListBucketsResponse): S3BucketListing =
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]: Make this a private helper in the implementation.

File: src/main/scala/zio/s3/S3.scala [hunk 23]
```diff
@@ -0,0 +1,395 @@
...
+    S3BucketListing(resp.buckets().asScala.toList.map(S3Bucket(_)))
+}
+
+case class S3Bucket(name: String, creationDate: Instant)
+
+object S3Bucket {
+  def apply(bucket: Bucket): S3Bucket = new S3Bucket(bucket.name(), bucket.creationDate())
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]: `fromBucket`?

File: src/main/scala/zio/s3/S3.scala [hunk 24]
```diff
@@ -0,0 +1,395 @@
...
+}
+
+case class S3ObjectListing(
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]: `final`

File: src/main/scala/zio/s3/S3.scala [hunk 25]
```diff
@@ -0,0 +1,395 @@
...
+  bucketName: String,
+  objectSummaries: List[S3ObjectSummary],
+  nextContinuationToken: Option[String]
+)
+
+object S3ObjectListing {
+
+  def apply(r: ListObjectsV2Response): S3ObjectListing =
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]: `fromXYZ`

File: src/main/scala/zio/s3/S3.scala [hunk 26]
```diff
@@ -0,0 +1,395 @@
...
+    S3ObjectListing(
+      r.name(),
+      r.contents().asScala.toList.map(o => S3ObjectSummary(r.name(), o.key())),
+      Option(r.nextContinuationToken())
+    )
+}
+
+case class S3ObjectSummary(bucketName: String, key: String)
```
Comment (jdegoes) [inline src/main/scala/zio/s3/S3.scala]: `final`

Comment (jdegoes) [reply src/main/scala/zio/s3/S3.scala]: Pull out all the `S3` things into `s3model.scala`

File: src/main/scala/zio/s3/errors.scala [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.s3
+
+case class ConnectionError(message: String, cause: Throwable) extends Exception(message, cause)
```
Comment (jdegoes) [inline src/main/scala/zio/s3/errors.scala]: `final` etc

File: src/main/scala/zio/s3/package.scala [hunk 28]
```diff
@@ -0,0 +1,107 @@
...
+package zio
+
+import java.util.concurrent.CompletableFuture
+
+import software.amazon.awssdk.services.s3.S3AsyncClient
+import software.amazon.awssdk.services.s3.model.S3Exception
+import zio.s3.ZIOStreamUtils._
+import zio.stream.{ ZSink, ZStream, ZStreamChunk }
+
+package object s3 extends S3.Service[S3] {
```
Comment (jdegoes) [inline src/main/scala/zio/s3/package.scala]: You will still have helpers but won't extend `S3.Service` anymore.

File: src/main/scala/zio/s3/settings.scala [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+package zio.s3
+
+import software.amazon.awssdk.regions.Region
+import zio.{ IO, ZIO }
+
+final case class S3Credentials(accessKeyId: String, secretAccessKey: String)
+
+final private[s3] case class S3Settings(region: Region, credentials: S3Credentials)
```
Comment (jdegoes) [inline src/main/scala/zio/s3/settings.scala]: Maybe a little wrapper for `Region`, a nice sum type?

**Comment by regis-leray** [comment]
@jdegoes i finally had time to finish s3 with all the feedbacks.

**Comment by jdegoes** [comment]
@regis-leray Awesome! Can you sign the contributor license's agreement, then I'll review and we can merge this on it?

**Comment by regis-leray** [comment]
@jdegoes  CLA is signed and ready to review

**Comment by jdegoes** [comment]
@regis-leray Great work!

**Pull request merged**